server/src/main/java/org/opensearch/identity/Subject.java

```diff
@@ -5,15 +5,17 @@
 
 package org.opensearch.identity;
 
-import org.opensearch.identity.tokens.AuthToken;
+import org.opensearch.common.annotation.ExperimentalApi;
 
 import java.security.Principal;
+import java.util.concurrent.Callable;
 
 /**
  * An individual, process, or device that causes information to flow among objects or change to the system state.
  *
  * @opensearch.experimental
  */
+@ExperimentalApi
 public interface Subject {
```

Comment:
I don't see the purpose to declare additional subject types. What is the utility?

Reply:
I created this to address an earlier comment. I can't find the original thread, but there is another related thread [here](https://github.com/opensearch-project/OpenSearch/pull/14630#discussion_r1720286212).

The problem was the `authenticate` method on the original `Subject` interface would not be applicable in this case with a `PluginSubject`. The plugins don't have an auth token to authenticate with. The PluginSubject is secure because its provided to IdentityAware plugins on bootstrap so it cannot be shared between the plugins. You can already consider it authenticated if you have an instance of the subject.

Reply:
In this PR, `authenticate` has been moved to the UserSubject.

Reply:
> The plugins don't have an auth token to authenticate with. 

While they don't have one today, nothing prevents us from coming up with a handshake for registering plugins at runtime... extensions needed that capability.

> The PluginSubject is secure because its provided to IdentityAware plugins on bootstrap so it cannot be shared between the plugins.

**Not secure.** `IdentityPlugin.getPluginSubject(...)` allows you to put in a plugin and get its identity out.  If you can get to the instance of the IdentityPlugin & PluginService you can assume the identity of other plugins.

A token provided by the plugin, maybe as part of the IdentityAwarePlugin would be needed to call a system like this secure.

----

Why does it matter why kind of Subject is passed to the plugin, it doesn't know the implementation, no?

Reply:
> If you can get to the instance of the IdentityPlugin & PluginService you can assume the identity of other plugins.

How would a malicious plugin get access to these? The core repo can hold an instance of PluginSubject, but IdentityAwarePlugins should only have access to their pluginSubject and not be able to obtain other PluginSubjects.

Reply:
> A token provided by the plugin, maybe as part of the IdentityAwarePlugin would be needed to call a system like this secure.

This is not how the extensions flow works. Extensions need to be configured by a cluster administrator. On bootstrap, core will iterate through the extensions and start the handshake process. At the end of the handshake, core would use the IdentityPlugin to pass a Service account token to an extension.

I see this process as analogous. Core is aware of the installed plugins and as part of the bootstrap, identity issues a PluginSubject that it then sends to an installed plugin through IdentityAwarePlugin.assignSubject

Reply:
> Not secure. IdentityPlugin.getPluginSubject(...) allows you to put in a plugin and get its identity out. If you can get to the instance of the IdentityPlugin & PluginService you can assume the identity of other plugins.

Hm ... wouldn't the same argument apply to `SecurityPlugin` & `PluginService`? I think if the system is wide open - nothing is secure. I see a couple of ways we could make sure `IdentityPlugin.getPluginSubject(...)` is called in the right context only (using `SecurityManager` fe) since only core could call it.